File: majority_element.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        #Boyer - Moore Major Vote Algorithm
        
        major, count = nums[0], 1
        for i in range(1, len(nums)):
            if count == 0:
                count += 1
                major = nums[i]
            elif major == nums[i]:
                count += 1
            else:
                count -= 1
        return major
    
    #normal sorting
    #return sorted(nums)[len(nums)//2]

File: majority_element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        if(nums.size() == 0) return -1;
        int candidate = 0;
        int count = 1;
        
        for(int i=0; i<nums.size(); i++) {
            if(nums[candidate] == nums[i]) count++;
            else count--;
            
            if(count == 0) { //this means that a new element is encountered
                candidate = i;
                count = 1;
            }
        }
        return nums[candidate];
    }
};
